File: crunchy-core/src/main/java/io/crunchy/core/pipeline/ProcessingContext.java
package io.crunchy.core.pipeline;

import io.crunchy.core.model.CompressedContext;
import io.crunchy.core.model.CompressionLevel;
import io.crunchy.core.model.NormalizedDocument;

import java.util.ArrayList;
import java.util.List;

/** Shared state passed through every pipeline stage. */
public final class ProcessingContext {

    /** The CIR list an extracted sentence belongs to. */
    public enum CirList {
        DECISIONS, CONSTRAINTS, ACCEPTANCE_CRITERIA, RISKS, TODOS, DEPENDENCIES
    }

    /**
     * One extracted sentence, tagged with the category that determines its
     * value score. Extraction records these instead of writing straight into
     * the result, so the assembly stage can gate them by compression level.
     */
    public record Extract(CirList list, String text, Block.Category category, Block block) {}

    private final NormalizedDocument document;
    private final CompressionLevel level;
    private final List<SourceText> sources = new ArrayList<>();
    private final List<Block> blocks = new ArrayList<>();
    private final CompressedContext result = new CompressedContext();
    private final List<String> ignoredContent = new ArrayList<>();
    private final List<Extract> extracts = new ArrayList<>();

    public ProcessingContext(NormalizedDocument document, CompressionLevel level) {
        this.document = document;
        this.level = level;
        sources.add(new SourceText("description", "", false, document.getDescription()));
        List<NormalizedDocument.Comment> comments = document.getComments();
        for (int i = 0; i < comments.size(); i++) {
            NormalizedDocument.Comment c = comments.get(i);
            sources.add(new SourceText("comment:" + (i + 1), c.author(), c.bot(), c.body()));
        }
    }

    public NormalizedDocument getDocument() { return document; }
    public CompressionLevel getLevel() { return level; }
    public List<SourceText> getSources() { return sources; }
    public List<Block> getBlocks() { return blocks; }
    public CompressedContext getResult() { return result; }
    public List<String> getIgnoredContent() { return ignoredContent; }

    public List<Extract> getExtracts() { return extracts; }

    /** Records an extracted sentence and tags its source block's category. */
    public void addExtract(CirList list, String text, Block.Category category, Block block) {
        block.getCategories().add(category);
        extracts.add(new Extract(list, text, category, block));
    }

    /** Blocks that survived filtering, deduplication and ranking. */
    public List<Block> liveBlocks() {
        return blocks.stream().filter(b -> !b.isDropped()).toList();
    }
}

File: web/src/core/pipeline/ProcessingContext.ts
import { CompressedContext } from '../model/CompressedContext.js';
import { CompressionLevel } from '../model/CompressionLevel.js';
import { NormalizedDocument } from '../model/NormalizedDocument.js';
import { Block, Category } from './Block.js';
import { SourceText } from './SourceText.js';

/** The CIR list an extracted sentence belongs to. */
export type CirListName =
  | 'decisions'
  | 'constraints'
  | 'acceptanceCriteria'
  | 'risks'
  | 'todos'
  | 'dependencies';

/**
 * One extracted sentence, tagged with the category that determines its value
 * score. Extraction records these instead of writing straight into the result,
 * so the assembly stage can gate them by compression level.
 */
export interface Extract {
  list: CirListName;
  text: string;
  category: Category;
  block: Block;
}

/** Shared state passed through every pipeline stage. */
export class ProcessingContext {
  readonly sources: SourceText[] = [];
  readonly blocks: Block[] = [];
  readonly result = new CompressedContext();
  readonly ignoredContent: string[] = [];
  readonly extracts: Extract[] = [];

  constructor(
    readonly document: NormalizedDocument,
    readonly level: CompressionLevel,
  ) {
    this.sources.push(new SourceText('description', '', false, document.description));
    document.comments.forEach((c, i) => {
      this.sources.push(new SourceText(`comment:${i + 1}`, c.author, c.bot, c.body));
    });
  }

  /** Records an extracted sentence and tags its source block's category. */
  addExtract(list: CirListName, text: string, category: Category, block: Block): void {
    block.categories.add(category);
    this.extracts.push({ list, text, category, block });
  }

  /** Blocks that survived filtering, deduplication and ranking. */
  liveBlocks(): Block[] {
    return this.blocks.filter((b) => !b.dropped);
  }
}
